added fixstars and azalt

rstub opened the pull request.

vreijs committed:
added fixstars and azalt

vreijs committed:
added fixstars and azalt

vreijs committed:
Merge branch 'master' of https://github.com/vreijs/swephR

rstub commented on src/swephR.cpp:
```diff
@@ -99,11 +100,51 @@ void set_delta_t_userdef (double delta_t) {
 //' @export
 // [[Rcpp::export(swe_calc)]]
-Rcpp::List calc(double tjd_et, int ipl, int iflag = 0) {
+Rcpp::List calc(double tjd_et, int ipl, int iflag = 4) {
```
Why do you change the default value to 4?

rstub commented on src/swephR.cpp:
```diff
@@ -99,11 +100,51 @@ void set_delta_t_userdef (double delta_t) {
+Rcpp::List fixstar(std::string star, double tjd_et, int iflag = 4) {
+  std::array<double, 6> xx;
+  std::array<char, 256> serr;
+  star = star + "                                         ";
```
Please change this to `star.resize(41);`.

rstub commented on tests/some-test-lines.R:
```diff
@@ -0,0 +1,20 @@
+swe_set_ephe_path("C:\\ARCHAEOCOSMO\\ephe")
```
This is problematic:
* `library(swephR)` is needed first
* the path is OS and environment dependent

Please keep in mind that code in the `test` directory is automatically executed during `R CMD check`

rstub reviewed (commented):
@vreijs I have added three comments in this PR. Can you please take a look?

vreijs commented:
Hello Ralf,

On 14 August 2018 at 22:12, Ralf Stubner <notifications@github.com> wrote

> -Rcpp::List calc(double tjd_et, int ipl, int iflag = 0) {
> +Rcpp::List calc(double tjd_et, int ipl, int iflag = 4) {
>
> Why do you change the default value to 4?
>
But why '0'? In some way I find the default value here a little difficult
(I use my default close to 38418). Perhaps don't give any default here! The
minimum property of the iflag is the type of ephemeris used. I find it
minimally Moshier (=4), but you could also see it as DE431 (=1). I think by
the way tha =0 is not defined in SE.


> ------------------------------
>
> In src/swephR.cpp
> <https://github.com/rstub/swephR/pull/8#discussion_r210104291>:
>
> +  star = star + "                                         ";
>
> Please change this to star.resize(41);.
>
But the statement I have is also ok, or not? At least that worked for me
some 10 years now in VBA. My minimum length is 41 characters (if 'star' has
no length...). But anyway, I now changed it to star.resize(41);



> ------------------------------
>
> In tests/some-test-lines.R
> <https://github.com/rstub/swephR/pull/8#discussion_r210104842>:
>
> > @@ -0,0 +1,20 @@
> +swe_set_ephe_path("C:\\ARCHAEOCOSMO\\ephe")
>
> This is problematic:
>
>    - library(swephR) is needed first
>
>
Understand

>
>    - the path is OS and environment dependent
>    Please keep in mind that code in the test directory is automatically
>    executed during R CMD check
>
> But how do I cater for this OS depending essential information? As my
ephemeris files (~3 GByte) are already in an existing directory and used by
other existing SE apparition (which don't want to be related to R depending
folders), I need to use that OS and user depending directory. But perhaps I
should not include it in the test, but then we can't test any SE functions
that needs ephemeris files (Mosher is thus handy;)... So a paradox and we
need to choice which application is most important...
As this moment i don't have a clear solution for the important point you
raise.

All the best,


Victor

vreijs committed:
added library sweph in tests

vreijs committed:
added resize

vreijs committed:
return instea dof rc in test

vreijs committed:
chnage the test stataments

vreijs committed:
chnage the test stataments

vreijs committed:
Update some-test-lines.R

vreijs committed:
Update some-test-lines.R

rstub commented on src/swephR.cpp:
```diff
@@ -124,7 +124,7 @@ Rcpp::List calc(double tjd_et, int ipl, int iflag = 4) {
   std::array<double, 6> xx;
   std::array<char, 256> serr;
-  star = star + "                                         ";
+  star = star.resize(41);
```
`std::string::resize` resizes a string in place. It does not return the resized string. So the correct syntax is

    star.resize(41)

without the assignment.

vreijs commented:
Perhaps another example of my paradoxial progress. I should have
encountered errors after making the change to 'star<-starr.rise(41)'; as I
did multiple 'Clean and Rebuild', 'Load All', etc. no error messages
yesterday. Now I did some one of them: lots of error message and they are
gone when removing the 'star<-'. So another example of inconsistencies. I
must do things fundamentally wrong. Perhaps I will remove all R programmes
from my laptop and start anew (including reboots).

I have also severe problems with getting pushing and pulling to work. I
might abandon trying GitHub version control for now (in two weeks time a
colleague of mine is back from holidays, I perhaps I will ask her for help.
I seem not able to understand how to push and pull I get several error
messages:
error when pushing:

>>> C:/Program Files/Git/bin/git.exe push origin refs/heads/masterTo https://github.com/vreijs/swephR
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'https://github.com/vreijs/swephR'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.


when pulling

>>> C:/Program Files/Git/bin/git.exe pullAuto-merging tests/some-test-lines.R
CONFLICT (content): Merge conflict in tests/some-test-lines.R
Automatic merge failed; fix conflicts and then commit the result.

And get these in my some-test-lines.R
<<<<<<< HEAD

test_that("Azimuth and altitude calculations", {
=======
test_that("Azimuth and altitude postions", {
>>>>>>> 50df4f75666e1bc3d7d8c54bcfc32b8607e55d35
  result <- swe_azalt(1234567,1,c(0,50,10),15,1013.25,c(186,22))
  expect_equal(result$xaz, c(114.636188,   8.210881,
8.209274),tolerance = .0000001)
})

I tried to remove the lines, no luck. I tried "git merge --abort" no
luck. I browsed, but not real help.


On 15 August 2018 at 06:24, Ralf Stubner <notifications@github.com> wrote:

> *@rstub* commented on this pull request.
> ------------------------------
>
> In src/swephR.cpp
> <https://github.com/rstub/swephR/pull/8#discussion_r210173598>:
>
> > @@ -124,7 +124,7 @@ Rcpp::List calc(double tjd_et, int ipl, int iflag = 4) {
>  Rcpp::List fixstar(std::string star, double tjd_et, int iflag = 4) {
>    std::array<double, 6> xx;
>    std::array<char, 256> serr;
> -  star = star + "                                         ";
> +  star = star.resize(41);
>
> std::string::resize resizes a string in place. It does not return the
> resized string. So the correct syntax is
>
> star.resize(41)
>
> without the assignment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/8#pullrequestreview-146333036>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FpLcvrlUpEBMOiyUNqmdAoDsjZVlks5uQ7CBgaJpZM4V9Kzc>
> .
>

vreijs committed:
error in fixstar

Merge branch 'master' of https://github.com/vreijs/swephR

# Conflicts:
#	tests/some-test-lines.R

vreijs commented:
So I made this change in swe_fixstar and now I can push again (while before
the change not...) I am really flabbergasted by this behavior. Any way I
think my latest version is in my fork...

On 15 August 2018 at 06:24, Ralf Stubner <notifications@github.com> wrote:

> *@rstub* commented on this pull request.
> ------------------------------
>
> In src/swephR.cpp
> <https://github.com/rstub/swephR/pull/8#discussion_r210173598>:
>
> > @@ -124,7 +124,7 @@ Rcpp::List calc(double tjd_et, int ipl, int iflag = 4) {
>  Rcpp::List fixstar(std::string star, double tjd_et, int iflag = 4) {
>    std::array<double, 6> xx;
>    std::array<char, 256> serr;
> -  star = star + "                                         ";
> +  star = star.resize(41);
>
> std::string::resize resizes a string in place. It does not return the
> resized string. So the correct syntax is
>
> star.resize(41)
>
> without the assignment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstub/swephR/pull/8#pullrequestreview-146333036>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AQa6FpLcvrlUpEBMOiyUNqmdAoDsjZVlks5uQ7CBgaJpZM4V9Kzc>
> .
>

vreijs committed:
added swe_rise_trans_true_hor function

vreijs committed:
rectified an error in \code argument

vreijs committed:
remove the default from the swe functions

rstub commented:
Hi Victor

I really don't understand where you problems with git and GitHub come from. There are some good tutorials outer there, e.g. http://swcarpentry.github.io/git-novice/. Maybe they )or the college you mention) can help.

But let's look at this PR first. As general rule, I will not merge until at least the automatic checks are fine. Currently it fails because you are using functions from the `testthat` package within your test file. I can onl assume that it fails on your computer as well when you press the "check" button in the "build" tab in RStudio. Or call `devtools::check()`at the R prompt.

The following should work:

1. Remove all calls to `test_that` from you test file leaving only the commands you want to execute, like you had in in the beginning. This way the file should run ok on your computer and "checking" the package should also succeed.  
2. Wrap the the entire code in 
```
 if(dir.exists("C:\\ARCHAEOCOSMO\\ephe")) {
  ...
}
```
That way it will only be executed when the directory exists like on your computer. Hopefully this will make the automated tests succeed. I can then merge the changes. 

For now, it will probably be easier if we go back to you sending me source updates via email, until you have sorted out your git problems. Even though this will highlight it even more, that you are able to spend more time on this project than me. Progress will be slower, but we will get there eventually.

cheerio
ralf

vreijs commented:
At this moment I am able to put everything on GitHub, it seems I found a
way to do it. I think when I do Push to close (in time) to each other the
system does not like that. I now do once in a while a Push and it works
(both for my forked repository swephR and ARCHAEOCOSMO. I did not yet look
at the upstream commands; I once used them and it became a mess, but I will
try again.

We will need to overcome in general this directory issue, as it is not over
come-able, IMHO. So I now put the ephemeris files and sefstars in
C:\sweph\ephe and use the statement swe_set_ephe_path(""). Reading swephprg
documentation this is a save solution. (from:
file:///C:/sweph/doc/swephprg.htm#_Toc505244914):



1.   check environment variable SE_EPHE_PATH; if it exists it is used, and
if it has invalid content, the program fails.

2.   Try to find the ephemeris files in the current working directory

3.   Try to find the ephemeris files in the directory where the executable
resides

4.   Try to find a directory named \SWEPH\EPHE in one of the following
three drives:

·         where the executable resides

·         current drive

·         drive C:

If I do Check Package (I now copied my tests in the test-misc.R file) I get:

==> devtools::test()
Loading swephR
Testing swephR
v | OK F W S | Contextv | 13       | test-misc.R

== Results ============================================================================
OK:       13
Failed:   0
Warnings: 0
Skipped:  0


So that looks ok..

Of course one coudl usesomethign like:

if(dir.exists("C:\\ARCHAEOCOSMO\\ephe")) {
  ...

But is it not good to see the test fail, because it can't file the
proper files to work? Of course we nee dot provide proper help to help
rectify it.


So I don't see a problem..

All the best,

Victor

P.S. I even was thinking why are the C files of SE in the
..\swephR\src\libswe directory in the project; why copy them while they are
already in the C:\sweph\src folder (the default directory of Swiss
Ephemeris)? Why not use that default directory? A reason could be that we
don't want to be exposed to changes in SE, while we did not adjust the R
wrapper for that. That is a reason enough, I agree.

vreijs committed:
move the test to the test.misc.R file and also remove an epxlicit ephermis folder (thus using SE's defualt methodlogy for this).

rstub commented on tests/some-test-lines.R:
```diff
@@ -1,28 +0,0 @@
-
-swe_set_ephe_path("C:\\ARCHAEOCOSMO\\ephe")
-swe_set_topo(0,50,10)
-test_that("Existing star position", {
```
This test still fails on the CI platforms, since there is no stars data there (yet). However, we can skip this test on these platforms. To do so, please add

    skip_on_travis()
    skip_on_appveyor()

to this test.

vreijs commented:
Why skip the test? It is an error if the star file is not there...
Don't understand.

rstub commented:
The star file will be part of the package soon, but it is not there yet. So on the machines that do the continuous integration for us, the star file is not present *right now*. By skipping the tests now and maybe one or two other changes, we should get this PR to "green". After that I can merge it and bring in the star file so we can remove these skip commands.

vreijs commented:
Understood. I will add these lines.

vreijs committed:
added skip_on line in test-misc.R

vreijs commented:
By the way: what about the Rp files that don't show for the package and the
expert interface?
And the errror message I get when starting a 'Clean and Rebuilt' or a
'Install and Restart':

Error in library.dynam.unload(name, system.file(package = name)) :
  DLL ‘swephR.dll’ was not loaded

Or don't you have this error? Or is that related to RStudio?


I pushed my changes for the skip_on lines to GitHub.
By the way I am not able to test if these two additional lines work: I
still have:

==> devtools::test()
Loading swephR
Testing swephR
v | OK F W S | Contextv | 13       | test-misc.R

== Results ============================================================================
OK:       13
Failed:   0
Warnings: 0
Skipped:  0

rstub commented:
I cannot reproduce these two errors. The problem with the library might be specific to RStudio, which I do not use here.

Anyway, we are almost there! The tests run fine, but "check" produces a WARNING. You should be able to reproduce this by running `devtools::check()` or pressing the appropriate button in RStudio. Now on Travis this WARNING is treated as failure, since one should strive for a package to produce zero ERRORs , WARNINGs and NOTEs, in particualr when one tries to submit to CRAN. In that case one will have to explain every one of them. And explaining that a WARNING is actually OK is (almost) impossible.

However, I will pass on this here an merge this PR this evening. After that I will make a few adjustments to have "check" work again and add the stars file, so that we can remove the skip statements ...

vreijs commented:
On 16 August 2018 at 15:56, Ralf Stubner <notifications@github.com> wrote:

> I cannot reproduce these two errors. The problem with the library might be
> specific to RStudio, which I do not use here.
>
> Anyway, we are almost there! The tests run fine, but "check" produces a
> WARNING. You should be able to reproduce this by running devtools::check()
> or pressing the appropriate button in RStudio.
>

I see no errors in my set-up. See the grab I did:

> devtools::test()Loading swephRTesting swephR√ | OK F W S | Context√ | 13       | test-misc.R [0.1 s]

== Results =====================================================================Duration:
0.2 s

OK:       13
Failed:   0
Warnings: 0
Skipped:  0


>
Perhaps I can get the error if I remove the directory \ephe in \sweph?
Success with the PR. Thanks for doing this work, much appreciated.

vreijs commented:
Great.

Added swe_deltat_ex and had again problems that this new function was not
working in the ARCHAEOCOSMO package after running library (sweph3).
swe_deltat_ex was visible in the available functions, but it did no execute:

> swe_deltat_ex(1234567,4)Error in swe_deltat_ex(1234567, 4) : object '_swephR_deltat_ex' not found


So back to swephR project.
It seems that the function was only visible in swephR project after a
roxygenise()! As soon as I did an Install, the NAMESCAPE files was reduced
(all the swe_* definition were gone)! So I deleted the folder swephR in the
'installed' folder (C:\Users\VRHome\Documents\R\win-library\3.5\swephR).
Then I did again an 'Install and Restart" in swephR project.
Now it seems the NAMESPACE file is more persistent in the installed
folder....
And the swe_deltat_ex can be executed in ARCHAOECOSMO.

Strange behavior, but perhaps I have a good workflow:
. delete swephR folder in the installed packages,
. do an Install and Restart
. perhaps more is needed, will work on that.

Related to this is perhaps that R installs Rtools a lot of times (while it
is on the system):
> devtools::load_all(".")
Loading swephR
Re-compiling swephR
trying URL 'https://cran.rstudio.com/bin/windows/Rtools/Rtools35.exe'
Content type 'application/x-msdos-program' length 108622512 bytes (103.6 MB)

Thisis another thing that happens regular (when doing >
devtools::load_all(".")):
[....]
** libs
C:/RBuildTools/3.5/mingw_32/bin/g++  -std=gnu++11
-I"C:/PROGRA~1/R/R-35~1.1/include" -DNDEBUG
-I"C:/Users/VRHome/Documents/R/win-library/3.5/Rcpp/include"        -O2
-Wall  -mtune=generic -UNDEBUG -Wall -pedantic -g -Og -c RcppExports.cpp -o
RcppExports.o
      3 [main] make (14184) C:\RBUILD~1\3.5\bin\make.exe: *** fatal error -
cygheap base mismatch detected - 0x2DE5410/0x612A5410.
When I do devtools::load_all(".")) again, no error happens.

It (R, RStudio and/or RTools) is all quite instable IMHO...

vreijs commented:
I might have a clue what causes these issues I have:. ARCHAEOCOSMO package
uses library(sweph3) while I also install a new sweph3 (in another session
of RStudio). I think, that ARCHAEOCOSMO blocks changes to the files in
sweph3 install folder, so swephR can't make changes... Will check in the
future that ARCHAEOCOSMO does not use swephR while debugging swephR. And
see if that stops the problems I have,

rstub commented:
Before you do any further work on this: I had to change a few things, e.g.

* "squash" all your commits into one 
* list the parameters for all functions in one place to avoid duplicates
* a new header that no longer requires us to specify function definitions

This mean s it will be difficult to merge your changes if you continue with your code base. Normally you would now update the master from upstream and create a new branch in your repository. But since you are not (yet) used to this workflow, I have come up with a different idea:

* I have invited you as a collaborator to this repository
* I have created a separate branch for you vreis/function1

Workflow:

* Clone this repository to your local machine
* Switch to "your" branch; in the "git" tab in RStudio you can change the branch
* Work normally within the branch
* After some time, I can merge the branch and create a new one for you based on the new master

BTW, I think part of your problems with git might be caused b using "amend previous commit" on a commit that you already pushed to GitHub. *Never* do that! It is fine to amend a commit that you have not yet pushed, but once it has been pushed, you must not change it.

vreijs commented:
Hello Ralf,

I understand to 'stop' for a moment and indeed learn/use the proper
workflow.
The problem is that I learn by doing (and making errors) and thus I need a
project where it is not a problem to make errors (or loose things;-). And I
am afraid with swephR about that!
So I want to trial this in a small environment. I might see if a colleague
of mine is willing to spend some time on this with me.

One thing that concerns me:
"BTW, I think part of your problems with git might be caused b using "amend
previous commit" on a commit that you already pushed to GitHub. *Never* do
that! It is fine to amend a commit that you have not yet pushed, but once
it has been pushed, you must not change it."
Is 'amend previous commit' a command. I never used that, I always use
'Commit' in RStudio.
It happens many times, IMHO, that one changes something changed earlier (I
also sometimes don't know what I changed earlier, so how do I keep track?).
I had hoped that the version control would track that.

rstub commented:
> I understand to 'stop' for a moment and indeed learn/use the proper workflow.

Misunderstanding. I am not asking you to stop until you learn the 'proper' workflow. I really don't want to loose the momentum we currently have. Instead I am suggesting a simplified workflow. Let's review the 'proper' workflow to see what I mean:

1. You create a fork of my repository.
2. You keep the master branch of your fork up-to-date by pulling from my repository.
3. You don't make 'normal' commits to the master branch of your fork
4. If you want to change something, you create a new branch based on the current master and work in there.
5. At one time you open a pull request for the branch
6. We discuss the changes, which might trigger changes by you
7. I merge the PR
8. You can delete the branch if you want to. Any new work goes to a new branch!

Let's see how these 8 steps come out in my simplified workflow:

1. not needed
2. not needed
3. The master branch in my repository is 'protected', so you cannot commit there
4. For now I will create the branches for you. You can then work in the branch as usual.
5. Either you or I can create the PR.
6. As above + I can also make changes if necessary
7. As above.
8. I will decide whether or not to keep the branch. Otherwise as above.

This simplified workflow frees you from the non-trivial task of keeping your fork up-to-date and allows me to provide simple fixes directly in the branch. I am often better at fixing something than explainign how to fix it. 

Does that sound workable to you?

> Is 'amend previous commit' a command. I never used that, I always use 'Commit' in RStudio. It happens many times, IMHO, that one changes something changed earlier (I also sometimes don't know what I changed earlier, so how do I keep track?). I had hoped that the version control would track that.

If you look at the "commit" window in RStudio, then there is a "amend previous commit" checkbox. That you used that is one possible explanation for the commit history in your repo:

```
* ad4c61c remove the default from the swe functions
* cab5c59 rectified an error in \code argument
* 56cae88 added swe_rise_trans_true_hor function
*   30b53de error in fixstar
|\  
| * 50df4f7 Update some-test-lines.R
| * f9aca00 Update some-test-lines.R
| * df40d26 chnage the test stataments
* | 638da3f chnage the test stataments
|/  
* 4808328 return instea dof rc in test
* b28bd68 added resize
* aa3320e added library sweph in tests
*   cd5b848 Merge branch 'master' of https://github.com/vreijs/swephR
|\  
| * 73606ed added fixstars and azalt
* | 85fec5c added fixstars and azalt
|/  
* 613b066 Delete libswe.a
* e091e38 Implement swe_calc
```

You can get a similar view with the "history" button in the "git" window in RStudio.

These forks are normal if multiple developers are involved. However, here it is only you. Since in both cases the two forking commits have identical commit messages (including identical typos!) I suspect that you committed something and pushed it to github. In the next commit you pressed "amend previous commit". When you tried to push that to github, you where refused and had to pull first, resulting in a merge commit. An alternative explanation would be that you have multiple working copies of the repository and made the same physical changes in both with independent commits. There are probably further  possibilities that I cannot think of right now ...

However, such messy commit histories can be cleaned up before merging it into master ...

vreijs commented:
Hello Ralf,

I indeed used 'Amend previous commit' (I was so naive that I thought it was
only to get the old comment there (so less typing;-)  I am not using it
anymore and no problems since yesterday.

By the way the file swephapi.h is much simplified! Great.
The @param need some standardisation I would say: we now used the SE names
(from the C-calls) but we could rename the parameters the swephR functions
use, so we have standardised names. What do you think?
<As some @param have the same or slightly different name, the full list now
is not that helpful (as one does not know which name is belonging to what
function)>

So step 4 is my first 'action' in the simplified workflow. If I go to your
branches: https://github.com/rstub/swephR/branches ,
I assume I need to click on the one that is highest in the list that has
'New pull)r request' (in this case ' vreijs/functions1
<https://github.com/rstub/swephR/tree/vreijs/functions1> Updated 2 days ago
by rstub <https://github.com/rstub>' ). Correct? I assume the ones that are
labelled 'Merged' or a red cross should not be used?


On 17 August 2018 at 13:00, Ralf Stubner <notifications@github.com> wrote:

>
>    1. You create a fork of my repository.
>    2. You keep the master branch of your fork up-to-date by pulling from
>    my repository.
>    3. You don't make 'normal' commits to the master branch of your fork
>    4. If you want to change something, you create a new branch based on
>    the current master and work in there.
>    5. At one time you open a pull request for the branch
>    6. We discuss the changes, which might trigger changes by you
>    7. I merge the PR
>    8. You can delete the branch if you want to. Any new work goes to a
>    new branch!
>
> Let's see how these 8 steps come out in my simplified workflow:
>
>    1. not needed
>    2. not needed
>    3. The master branch in my repository is 'protected', so you cannot
>    commit there
>    4. For now I will create the branches for you. You can then work in
>    the branch as usual.
>    5. Either you or I can create the PR.
>    6. As above + I can also make changes if necessary
>    7. As above.
>    8. I will decide whether or not to keep the branch. Otherwise as above.
>
>
